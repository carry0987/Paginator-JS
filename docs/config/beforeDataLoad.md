---
id: beforeDataLoad
title: beforeDataLoad
---

`beforeDataLoad` is a callback function that will be called before data is loaded. This function can be used to show a loading message or spinner.

-   `optional`
-   Type: `() => void`

```ts
new Paginator({
    // ...
    beforeDataLoad: () => {
        document.querySelector('.data-container').innerHTML = 'Loading data ...';
    }
});
```
