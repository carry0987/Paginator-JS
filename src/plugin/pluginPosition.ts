export enum PluginPosition {
    Header,
    Body,
    Footer,
    Cell
}
