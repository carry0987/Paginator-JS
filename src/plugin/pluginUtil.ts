import Header from '@/component/header';
import { classJoin, className } from '@/module/utils/className';

class PluginUtil {
    // Header
    public leafColumns = Header.leafColumns;

    // Style
    public classJoin = classJoin;
    public className = className;
}

export const pluginUtil = new PluginUtil();
