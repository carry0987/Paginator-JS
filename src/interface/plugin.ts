import { PluginPosition } from '@/plugin/pluginPosition';
import { FunctionComponent } from 'preact';

export interface Plugin<T extends FunctionComponent<P>, P = any> {
    id: string;
    position: PluginPosition;
    component: T;
    props?: P;
    order?: number;
}
