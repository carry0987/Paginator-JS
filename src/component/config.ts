import Header from './header';
import { Status } from '@/type/types';
import { Options } from '@/interface/options';
import { State } from '@/interface/state';
import Utils from '@/module/utils/utils-ext';
import { StateManager } from '@/module/state/stateManager';
import StorageUtils from '@/module/storage/storageUtils';
import PipelineUtils from '@/module/pipeline/pipelineUtils';
import { Translator } from '@/module/i18n/translator';
import { createContext } from 'preact';

class Config {
    public options: Options = {} as Options;

    constructor() {
        this.assign(Config.defaultConfig());
    }

    public assign(partialConfig: Partial<Options>): this {
        Utils.shallowMerge(this.options, partialConfig);

        return this;
    }

    public update(partialConfig: Partial<Options>): this {
        if (!partialConfig) return this;

        this.assign(
            Config.fromPartialConfig({
                ...this.options,
                ...partialConfig,
            }),
        );

        return this;
    }

    private static defaultConfig(): Partial<Options> {
        return {
            state: new StateManager<State>({
                status: Status.Init,
                tabular: null,
            }),
            position: 'bottom',
            resetPageOnUpdate: false,
            pageNumber: 1,
            pageSize: 10,
            pageRange: 2,
            display: {
                showPrevious: true,
                showNext: true,
                showPageNumbers: true,
                hideFirstOnEllipsisShow: false,
                hideLastOnEllipsisShow: false,
                autoHidePrevious: false,
                autoHideNext: false,
            },
            className: {
                container: '',
                active: 'active',
                disable: 'disabled',
                pageList: 'pages',
                pageButton: 'page-item',
                prevButton: 'page-prev',
                nextButton: 'page-next',
            },
        };
    }

    private static fromPartialConfig(partialConfig: Partial<Options>) {
        const config = new Config().assign(partialConfig);

        config.assign({
            header: Header.createFromConfig(config),
        });

        config.assign({
            storage: StorageUtils.createFromConfig(config),
        });

        config.assign({
            pipeline: PipelineUtils.createFromConfig(config),
        });

        config.assign({
            translator: new Translator(config.options.language),
        });

        return config.options;
    }
}

const ConfigContext = createContext<Config | undefined>(undefined);

export { Config, ConfigContext };
