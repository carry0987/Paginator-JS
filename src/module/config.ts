import { Options } from '../interface/interfaces';

export const defaults: Options = {
    dataSource: [],
    locator: 'data',
    totalNumber: 0,
    totalNumberLocator: null,
    pageNumber: 1,
    pageSize: 10,
    pageRange: 2,
    alias: {
        pageNumber: 'pageNumber',
        pageSize: 'pageSize',
    },
    showPrevious: true,
    showNext: true,
    showPageNumbers: true,
    showNavigator: false,
    hideFirstOnEllipsisShow: false,
    hideLastOnEllipsisShow: false,
    autoHidePrevious: false,
    autoHideNext: false,
    classPrefix: 'pagination',
    className: '',
    activeClassName: 'active',
    disableClassName: 'disabled',
    ulClassName: '',
    pageClassName: '',
    prevClassName: '',
    nextClassName: '',
    prevText: '&laquo;',
    nextText: '&raquo;',
    ellipsisText: '...',
    goButtonText: 'Go',
    formatNavigator: '{currentPage} of {totalPage}',
    header: '',
    footer: '',
    pageLink: '',
    position: 'bottom',
    formatResult: (data: any) => data,
    dataLoader: {
        method: 'GET',
        data: {},
        cache: 'no-cache',
        headers: {
            'Content-Type': 'application/json'
        },
        mode: 'cors',
        beforeSend: (pagination: any) => pagination,
        credentials: 'same-origin',
        pageNumberStartWithZero: false
    },
    triggerPagingOnInit: true,
    resetPageNumberOnInit: false,
    hideOnlyOnePage: false,
    onError: (err: unknown) => console.error('Pagination error:', err)
};
