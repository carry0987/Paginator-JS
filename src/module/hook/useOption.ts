import { Options } from '@/interface/options';
import { ConfigContext } from '@/component/config';
import { useContext } from 'preact/hooks';

export const useOption = (): Options => {
    const context = useContext(ConfigContext);

    if (context === undefined) {
        throw new Error('useOption must be used within a ConfigProvider');
    }

    return context.options;
};
