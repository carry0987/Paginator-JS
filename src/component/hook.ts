import { useStore } from '@/module/hook/useStore';
import { useSelector } from '@/module/hook/useSelector';
import { useConfig } from '@/module/hook/useConfig';
import { useOption } from '@/module/hook/useOption';
import { useTranslator } from '@/module/hook/useTranslator';

class Hook {
    static useStore: typeof useStore = useStore;
    static useSelector: typeof useSelector = useSelector;
    static useConfig: typeof useConfig = useConfig;
    static useOption: typeof useOption = useOption;
    static useTranslator: typeof useTranslator = useTranslator;
}

export default Hook;
