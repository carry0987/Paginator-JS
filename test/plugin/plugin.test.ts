import { Config, ConfigContext } from '@/component/config';
import { useOption } from '@/module/hook/useOption';
import PluginManager from '@/plugin/pluginManager';
import { PluginUtil } from '@/plugin/pluginUtil';
import { PluginPosition } from '@/plugin/pluginPosition';
import { describe, it, expect } from 'vitest';
import { Fragment, h, render } from 'preact';

interface DummyConfig extends Config {
    dummy: {
        text: string;
    };
}

interface TestProps {
    test: string;
}

describe('Plugin', () => {
    function DummyPlugin<T extends DummyConfig>() {
        const option = useOption() as unknown as T;

        return h('b', {}, option?.dummy?.text || 'hello!');
    }

    it('should add and remove plugins', () => {
        const manager = new PluginManager();

        expect(manager.list()).toHaveLength(0);

        manager.add<TestProps>({
            id: 'dummy',
            position: PluginPosition.Header,
            component: DummyPlugin,
            props: {
                test: 'test'
            }
        });

        manager.add({
            id: 'dummy2',
            position: PluginPosition.Header,
            component: DummyPlugin
        });

        expect(manager.list()).toHaveLength(2);
        manager.remove('dummy');
        expect(manager.list()).toHaveLength(1);
        manager.remove('dummy2');
        expect(manager.list()).toHaveLength(0);
        manager.remove('doesntexist');
        expect(manager.list()).toHaveLength(0);
    });

    it('should return unordered plugins', () => {
        const manager = new PluginManager();

        expect(manager.list()).toHaveLength(0);

        manager.add({
            id: 'dummy',
            position: PluginPosition.Header,
            component: DummyPlugin.prototype
        });

        manager.add({
            id: 'dummy2',
            order: 1,
            position: PluginPosition.Header,
            component: DummyPlugin.prototype
        });

        manager.add({
            id: 'dummy3',
            order: 10,
            position: PluginPosition.Footer,
            component: DummyPlugin.prototype
        });

        expect(manager.list().map((x) => x.id)).toStrictEqual(['dummy', 'dummy2', 'dummy3']);
    });

    it('should return plugins in the correct order', () => {
        const manager = new PluginManager();

        expect(manager.list()).toHaveLength(0);

        manager.add({
            id: 'dummy',
            order: 5,
            position: PluginPosition.Header,
            component: DummyPlugin.prototype
        });

        manager.add({
            id: 'dummy2',
            order: 1,
            position: PluginPosition.Header,
            component: DummyPlugin.prototype
        });

        manager.add({
            id: 'dummy3',
            order: 10,
            position: PluginPosition.Footer,
            component: DummyPlugin.prototype
        });

        manager.add({
            id: 'dummy4',
            position: PluginPosition.Footer,
            component: DummyPlugin.prototype
        });

        expect(manager.list().map((x) => x.id)).toStrictEqual(['dummy2', 'dummy', 'dummy3', 'dummy4']);
        expect(manager.list(PluginPosition.Header).map((x) => x.id)).toStrictEqual(['dummy2', 'dummy']);
    });

    it('should return existing plugin by id', () => {
        const manager = new PluginManager();

        const component = DummyPlugin.prototype;
        manager.add({
            id: 'dummy',
            position: PluginPosition.Header,
            component: component
        });

        const plugin = manager.get('dummy');

        expect(plugin!.component).toBe(component);
        expect(plugin!.position).toBe(PluginPosition.Header);

        expect(manager.get('doesnexist')).toBeUndefined();
    });

    it('should create a userConfig with custom plugin', () => {
        const config = new Config().update({
            data: [[1, 2, 3]],
            plugins: [
                {
                    id: 'dummyheader',
                    position: PluginPosition.Header,
                    component: DummyPlugin
                }
            ]
        });

        expect(config.internal.plugin.get('dummyheader')).toMatchObject({
            id: 'dummyheader',
            position: PluginPosition.Header
        });
    });

    describe('pluginUtil', () => {
        describe('pluginRenderer', () => {
            it('should render the plugins', async () => {
                const config = new Config().update({
                    data: [[1, 2, 3]]
                }) as DummyConfig;

                (config.options as any).dummy = {
                    text: 'dummyplugin'
                };

                config.internal.plugin.add({
                    id: 'dummyheader',
                    position: PluginPosition.Header,
                    component: DummyPlugin
                });

                config.internal.plugin.add({
                    id: 'dummyfooter',
                    position: PluginPosition.Footer,
                    component: DummyPlugin
                });

                const container = document.createElement('div');
                document.body.appendChild(container);

                render(
                    h(
                        ConfigContext.Provider,
                        { value: config },
                        h(
                            Fragment,
                            null,
                            h(PluginUtil.PluginRenderer, { position: PluginPosition.Header }),
                            h(PluginUtil.PluginRenderer, { position: PluginPosition.Footer })
                        )
                    ),
                    container
                );

                expect(container.innerHTML).toMatchSnapshot();
            });
        });

        describe('className', () => {
            it('should accept two or more args', () => {
                expect(PluginUtil.className('boo', 'foo', 'bar')).toBe('paginatorjs-plugin-boo-foo-bar');
            });

            it('should generate classNames', () => {
                expect(PluginUtil.className('boo')).toBe('paginatorjs-plugin-boo');
            });
        });
    });
});
