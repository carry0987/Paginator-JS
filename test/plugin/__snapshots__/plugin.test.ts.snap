// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Plugin > should render the plugins 1`] = `"<b>hello!</b><b>hello!</b>"`;
